stdlib/either.mli

```diff
@@ -0,0 +1,88 @@
+(*   Copyright 2019 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Either type.
+
+    Either is the simplest and most generic sum type. It is isomorphic
+    to the [result] type; it should be used when there is no implied
+    success/failure semantics in the choice of one of the constructors.
+
+    @since 4.11 *)
+
+(* Unlike [result], no [either] type is made available in Stdlib,
+   one needs to access [Either.t] explicitly:
+
+   - This type is less common in typical OCaml codebases,
+     which prefer domain-specific variant types whose constructors
+     carry more meaning.
+   - Adding this to Stdlib would raise warnings in existing codebases
+     that already use a constructor named Left or Right:
+     + when opening a module that exports such a name,
+       warning 45 is raised
+     + adding a second constructor of the same name in scope kicks
+       in the disambiguation mechanisms, and warning 41 may now
+       be raised by existing code.
+
+   If the use becomes more common in the future we can always
+   revisit this choice.
+*)
+
+type ('a, 'b) t = Left of 'a | Right of 'b (**)
+(** A value of [('a, 'b) Either.t] contains either a value of ['a]  or a value
+    of ['b] *)
+
+val left : 'a -> ('a, 'b) t
+(** [left v] is [Left v]. *)
+
+val right : 'b -> ('a, 'b) t
+(** [right v] is [Right v]. *)
+
+val get_left : ('a, 'b) t -> 'a option
+(** [get_left (Left v)] is [Some v], [get_left (Right _)] is [None] *)
+
+val get_right : ('a, 'b) t -> 'b option
+(** [get_right (Right v)] is [Some v], [get_right (Left _)] is [None] *)
+
+val swap : ('a, 'b) t -> ('b, 'a) t
+(** [swap (Left v)] is [Right v], and [swap (Right v)] is [Left v]. *)
+
+val map_left : ('a1 -> 'a2) -> ('a1, 'b) t -> ('a2, 'b) t
+(** [map_left f e] is [Left (f v)] if [e] is [Left v]
+    and [e] if [e] is [Right _]. *)
+
+val map_right : ('b1 -> 'b2) -> ('a, 'b1) t -> ('a, 'b2) t
+(** [map_right f e] is [Right (f v)] if [e] is [Right v] and [e] if [e] is [Left _]. *)
+
+val map_both : left:('a1 -> 'a2) -> right:('b1 -> 'b2) -> ('a1, 'b1) t -> ('a2, 'b2) t
+(** [map_both ~left ~right e] is [Left (left v)] if [e] is [Left v]
+    and [Right (right v)] if [e] is [Right v]. *)
+
+val fold : left:('a -> 'c) -> right:('b -> 'c) -> ('a, 'b) t -> 'c
+(** [fold ~left ~right (Left v)] is [left v], and
+    [fold ~left ~right (Right v)] is [right v]. *)
+
```

Comment:
It would be nice to add predicates `is_left` and `is_right` to mirror `is_ok` and `is_error`.

Reply:
I started with them (I just copied your API from result), but then I got rid of them. I'm not fully convinced they are better than `(function Left _ -> true | Right _ -> false)`, and I'm not sure we should encourage the style of "if Either.is_left ... then ... else ...", which I find inelegant.

Reply:
As a point of reference: they appear in the corresponding Haskell module.

Reply:
@gasche the point is not to write `if Either.is_left ... then ... else ...`. The point is to be able to write `List.filter Either.is_left` instead of `List.filter (function Left _ -> true | Right _ -> false)`.

Reply:
I'm not sure about this: is it useful to build a list of `Left` values (`[Left x; Left y; Left z; ...]`)?  I'd expect `List.filter_map Either.get_left` (which will build `[x; y; z; ...]` instead) to be used more often.

Reply:
@yallop yes it is ! Here are few arguments.

1. You may be interested in filtering things before adding more eithers to a list or in general to a filterable data structure.
2. You may want to check everything is left or right but still keep it in either, so you want to write `List.for_all Either.is_left` rather than `List.forall (function Left _ -> true | Right _ -> false)`.
3. We recently introduced `Option`, `Result` whose interfaces were reviewed to death. I think now `Either` can afford consistency with those without us spending days in discussing it (both `Result.is_{ok,error}`, `Option.is_{none,some}` do exist).

Reply:
I think that `List.filter Either.is_left` is bad design, because it returns a list that you know contains only `'a` (starting from `('a, 'b) Either.t`), but you cannot actually get those `'a` without using an unsafe/partial function. You should use `List.filter_map Either.find_left` instead.

I'm not convinced by `List.for_all Either.is_left` either. The point of having a list of either values is that both cases can occur in it. If you're creating a value that at some point is known by construction to have only `Left (v : 'a)` elements, why not create a list of `'a` directly at this point? If really the way the postcondition holds is too complex to be built into the types in this way, well this sounds infrequent enough that just using `match ... ` or `function ...` is fine.

I'm still looking for a use-case of `is_left` that I personally find convincing. I'm not in favor of adding the function unless one is provided.

(I did not participate much to the review of the `Option` and `Result` type -- and in any case I wanted the types in more than I cared to debate detail points of the API -- but also there I don't think that those are good functions to have. Not my problem, and not my PR.)

Reply:
Who are you to judge what I want to do with these lists ? Again I may want to keep lefts or rights from from a particular list to compose it with another list of eithers. 

Besides regardless of whether you did participate or not to the `Option` and `Result` PRs, stdlib PR are also judged on consistency of the stdlib. By leaving these functions out you are making it less consistent. 

Frankly I think if you simply had taken these two recent modules as a complete design blueprint (and maybe added `find_{left,right}`) we would already be done at that point.

Reply:
If you use `List.filter_map Either.find_left`, you can still merge the list with other either values if you wish by making a `List.map Either.left` first. I suspect that in most cases the code will be clearer that way. Of course there are always some odd situations where unnatural code patterns emerge, but I think in this case it's fine to have less stdlib support -- especially if making those unnatural code patterns easier also risks nudging users towards worse decisions in more common cases.

Can you or someone else suggest a *compelling* use-case of these `is_*` functions?

I tried to use github search a bit, and most uses of `Option.is_some` are in code where pattern-matching would be better (see for example [here](https://github.com/ngorogiannis/cyclist/blob/6d4ad7ac5baa7286f376a4726baa8495e393cadd/src/util/tags.ml)). `is_{ok,error}` seems to have some slightly more believable use-cases (here are the [search results for Result.is_error](https://github.com/search?p=2&q=Result.is_error++extension%3Aml&type=Code)), mostly in testing code. But I don't think that these uses (related to the fact that it's typically enough to check if the result is Error to know that a process failed) would port well to Either use-cases.

Reply:
> Can you or someone else suggest a _compelling_ use-case of these `is_*` functions?

Stream source merges/split/re-source. Test suites. Whatever. Anything that *your* brain may not have imagined of the form: I want to take the lefts of that data structure and put them in the eithers of that other data structure. For that I'm sorry but, `map Either.left (filter_map Either.find_left)` is *not* clearer than `filter Either.is_left` (and I won't even talk about allocation this discussion is getting absurd).

I can't believe you are proposing an inconsistent design (including the way you initially called the functions you now renamed to `find_`) just one year after we collectively agreed on one (FWIW I was not ok with all of what when into `Option` and `Result` modules). I think this is just disrespectful for the work that has been done and a waste of (at least my) time.